### PR TITLE
API/UI: add nvim_input_mouse() including multigrid mouse support

### PR DESCRIPTION
**Pull request by bfredl**
Needs more tests (and likely edge cases to fix). Also docs, as far as I can see the existing `<LeftMouse><2,1>` format hasn't been properly documented, should probably go in `ui.txt`.

ref #9421

### CODE REVIEW

**Review by justinmk — APPROVED** [review]

**Comment by bfredl** [comment]
Instead of extending the `nvim_input("<LeftMouse><col,row>")` format, should we introduce `nvim_mouse_input(button,event,grid,row,col)` API method, and deprecate the stringly format? It seems `nvim_input("<LeftMouse><20,6><LeftDrag><20,5>")` doesn't work (need two separate calls), so these do not work like ordinary keys anyway.

**Comment by bfredl** [comment]
Added modifier support (and tests), it should be feature complete.

**Comment by bfredl** [comment]
While working on this I identified that `terminal_resize` is not used correctly with multigrid. I will factor it in a separate PR tough, as it looks a bit complex. (like: why does it both take width as parameter and try to calculate width from windows at the same time? what does the `terminal_resize(wp, 0, 0)` call do, nothing?)

**Comment by bfredl** [comment]
Fixed the remaining error (calculate cursor pos after vertical resize).

File: src/nvim/api/vim.c [hunk 1]
```diff
@@ -207,6 +211,98 @@ Integer nvim_input(String keys)
   return (Integer)input_enqueue(keys);
 }
 
+/// Send mouse event from GUI
+///
+/// The call is non-blocking. It doesn't wait on any resulting action, but
+/// queues the event to be processed soon by the event loop.
+///
+/// @note Currently this doesn't not support "scripting" multiple mouse events
```
Comment (justinmk) [inline src/nvim/api/vim.c]:
```suggestion
/// @note Currently this doesn't support "scripting" multiple mouse events
```

File: src/nvim/api/vim.c [hunk 2]
```diff
@@ -207,6 +211,98 @@ Integer nvim_input(String keys)
...
+///       by calling it multiple times in a loop: the intermediate mouse
+///       positions will be ignored. It should be used to implement real-time
+///       mouse input in a GUI. The deprecated mouse support in |nvim_input()|
```
Comment (justinmk) [inline src/nvim/api/vim.c]:
```suggestion
///       mouse input in a GUI. The deprecated pseudokey form (e.g. "<LeftMouse><col,row>")
```

File: test/functional/ui/mouse_spec.lua [hunk 3]
```diff
@@ -617,7 +635,7 @@ describe('ui/mouse/input', function()
     feed('<cr>')
   end)
 
-  it('mouse whell will target the hovered window', function()
+  local function wheel(use_meth)
```
Comment (justinmk) [inline test/functional/ui/mouse_spec.lua]: `use_api` ? Would like to move away from referring to the API as "methods", I don't think that makes sense to newcomers.

File: test/functional/ui/mouse_spec.lua [hunk 4]
```diff
@@ -698,6 +729,14 @@ describe('ui/mouse/input', function()
       {4:[No Name] [+]                                        }|
       :vsp                                                 |
     ]])
+  end
+
+  it('mouse wheel will target the hovered window', function()
+    --wheel(false)
```
Comment (justinmk) [inline test/functional/ui/mouse_spec.lua]: is this intentional?

File: test/functional/ui/mouse_spec.lua [hunk 5]
```diff
@@ -732,6 +771,39 @@ describe('ui/mouse/input', function()
     ]])
   end)
 
+  it('horizontal scrolling (input_mouse)', function()
```
Comment (justinmk) [inline test/functional/ui/mouse_spec.lua]:
```suggestion
  it('horizontal scrolling (nvim_input_mouse)', function()
```

just like we don't abbreviate option/commands, we should avoid abbreviating API function names, particularly in test descriptions.

File: src/nvim/api/vim.c [hunk 6]
```diff
@@ -207,6 +211,98 @@ Integer nvim_input(String keys)
...
+/// @param button Which mouse button, one of "left", "right", "middle" or
+///               "wheel".
+/// @param action For ordinary buttons, one of "press", "drag" and "release"
+///               For the wheel, use "up", "down", "left" and "right".
+/// @param modifier String of modifiers each represented by a single char.
+///                 The same specifiers are used as for a key press, except
+///                 that the "-" separator is optional, so "C-A-", "c-a"
+///                 and "CA" can all be used to specify Ctrl+Alt+click
+/// @param grid For a client using |ui-multigrid| pass in the grid number.
+///             Other clients should pass in 0 (not 1).
```
Comment (justinmk) [inline src/nvim/api/vim.c]: what happens if a `ext_multigrid` client sends 0?

**Review by justinmk — APPROVED** [review]

**Comment by justinmk** [comment]
> While working on this I identified that `terminal_resize` is not used correctly with multigrid. I will fix/refactor it in a separate PR tough, as it looks a bit complex. (like: why does it both take width as parameter and try to calculate width from windows at the same time? what does the `terminal_resize(wp, 0, 0)` call do, nothing?)

`terminal_resize` is in a weird state. There is at least one pending PR for it. I've also been meaning to revisit it.

**Comment by bfredl** [comment]
Thanks, updated.

**Pull request merged**